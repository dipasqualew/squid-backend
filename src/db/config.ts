/* eslint-disable @typescript-eslint/no-var-requires */
import knex from "knex";
import type { DefaultContext } from 'koa';


const knexfile = require('../../knexfile');


export enum Tables {
  users = 'users',
  squids = 'squids',
  squids_contents = 'squids_contents',
  squads = 'squads',
}

export const TablePriority = [
  Tables.users,
  Tables.squids,
  Tables.squids_contents,
  Tables.squads,
];


export const getDB = (): knex => {
  return knex(knexfile);
};

export const DatabaseMiddleware = async (context: DefaultContext, next: () => Promise<void>): Promise<void> => {
  context.db = getDB();
  await next();
};
