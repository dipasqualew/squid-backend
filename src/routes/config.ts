import type { Method } from 'axios';

export interface Route {
  method: Method;
  path: string;
  label: string;
}

export const route = (method: Method, path: string): Route => ({
  method,
  path,
  label: `${method} ${path}`,
});

export const ROUTES = {
  SQUIDS__DELETE: route('DELETE', '/v1/squids/:uuid'),
  SQUIDS__DETAIL: route('GET', '/v1/squids/:uuid'),
  SQUIDS__LIST: route('GET', '/v1/squids'),
  SQUIDS__PUT: route('PUT', '/v1/squids'),
  STATUS: route('GET', '/v1/status'),
  USERS__AUTH: route('POST', '/v1/users/auth'),
};
