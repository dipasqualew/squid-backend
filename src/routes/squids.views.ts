import Router from '@koa/router';

import type { SquidAppState, SquidAppContext } from '../server';
import type { Squid } from '../db/models/Squid';

export const SquidDetail = async (context: SquidAppContext): Promise<void> => {
  const uuid: string = context.params.uuid;

  const squid = await context.db
    .select<Squid[]>('*')
    .from('squids')
    .where({ uuid })
    .first();

  if (squid) {
    context.status = 200;
    context.body = { data: squid };
  } else {
    context.status = 404;
    context.body = { data: null };
  }
};

export const SquidList = async (context: SquidAppContext): Promise<void> => {
  const squids = await context.db
    .select<Squid[]>('*')
    .from('squids');

  context.body = { data: squids };
};

export const SquidsController = new Router<SquidAppState, SquidAppContext>()
  .get('/v1/squids/:uuid', SquidDetail)
  .get('/v1/squids', SquidList);
