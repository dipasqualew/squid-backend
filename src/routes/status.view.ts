import Router from '@koa/router';

import type { SquidAppState, SquidAppContext } from '../server';

export const Status = async (context: SquidAppContext): Promise<void> => {
  context.body = { status: 'OK' };
};

export const V1_STATUS_VIEWS = new Router<SquidAppState, SquidAppContext>();

V1_STATUS_VIEWS
  .get('/v1/status', Status);
