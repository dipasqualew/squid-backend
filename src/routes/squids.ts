import Router from '@koa/router';

import { ROUTES } from './config';
import type { SquidAppState, SquidAppContext } from '../server';
import type { Squid } from '../db/models/Squid';

export const SquidsDetail = async (context: SquidAppContext): Promise<void> => {
  const uuid: string = context.params.uuid;

  const squid = await context.db
    .select<Squid[]>('squids.uuid', 'squids.owner_uuid', 'squids.title', 'squids.content_type', 'squids.public', 'squids_contents.contents')
    .from('squids')
    .leftJoin('squids_contents', 'squids.uuid', 'squids_contents.uuid')
    .where({ 'squids.uuid': uuid })
    .first();

  if (squid) {
    context.status = 200;
    context.body = { data: squid };
  } else {
    context.status = 404;
    context.body = { data: null };
  }
};

export const SquidsList = async (context: SquidAppContext): Promise<void> => {
  const squids = await context.db
    .select<Squid[]>('*')
    .from('squids');

  context.body = { data: squids };
};

export const SquidsController = new Router<SquidAppState, SquidAppContext>()
  .get(ROUTES.SQUIDS__DETAIL.path, SquidsDetail)
  .get(ROUTES.SQUIDS__LIST.path, SquidsList);
