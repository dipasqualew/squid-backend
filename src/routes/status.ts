import Router from '@koa/router';

import { ROUTES } from './config';
import type { SquidAppState, SquidAppContext } from '../server';

export const Status = async (context: SquidAppContext): Promise<void> => {
  context.body = { status: 'OK' };
};

export const StatusController = new Router<SquidAppState, SquidAppContext>();

StatusController
  .get(ROUTES.STATUS.path, Status);
