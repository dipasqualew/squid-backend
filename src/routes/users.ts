import Router from '@koa/router';
import type { Context } from 'koa';

import { Claim, sign } from '../auth';
import { getDB } from '../db/config';
import { ROUTES } from './config';


export const UsersAuth = async (context: Context): Promise<void> => {
  const db = getDB();
  const rows = await db
    .select<Claim[]>('uuid', 'email')
    .from('users')
    .where({
      email: context.request.body.email,
      password: db.raw(`crypt('${context.request.body.password}', password)`),
    });

  if (rows.length === 1) {
    const jwt = await sign(rows[0]);

    context.response.status = 200;
    context.response.body = { jwt };
  } else {
    context.response.status = 403;
    context.response.body = {
      errors: [
        'Not Authorized',
      ],
    };
  }
};

export const UsersController = new Router()
  .post(ROUTES.USERS__AUTH.path, UsersAuth);
