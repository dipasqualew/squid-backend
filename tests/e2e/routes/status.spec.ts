import { AxiosResponse } from 'axios';

import { ROUTES } from '../../../src/routes/config';
import { RouteTester } from '../utils';

describe('e2e.routes.status', () => {
  RouteTester.test(ROUTES.STATUS, (tester) => {
    let response: AxiosResponse | null = null;

    tester.beforeAll(async () => {
      response = await tester.query();
    });

    tester.afterAll();

    it('returns HTTP 200', () => {
      expect(response?.status).toEqual(200);
    });

    it('returns the application state', () => {
      expect(response?.data).toMatchSnapshot();
    });
  });
});
