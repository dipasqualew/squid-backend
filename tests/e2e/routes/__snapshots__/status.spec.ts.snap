// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e.routes.status GET /v1/status returns the application state 1`] = `
Object {
  "status": "OK",
}
`;
